### PR TITLE
[Popover] Fix issue with Popover not being openable in a small Scrollable container

### LINKED ISSUE #544: Popover doesn't open in Modal
# Issue summary

When a `Popover` is used inside of a `Modal`, the popover never activates in the "desktop" rendered version of the mobile. It works in the "mobile" version.

## Expected behavior

Interacting with the activator should cause the `Popover` content to show.

## Actual behavior

Interacting with the activator for the `Popover` does not trigger the popover content to show.

This does work when the modal is in its mobile rendering, as it tends to have enough space in its `Scrollable` container, but in the normal desktop rendering, the modal's `Scrollable` seems to be triggering the scrollout on the `Popover`.

Image of issue

## Steps to reproduce the problem

1. Place a `Popover` component inside of a `Modal`
1. Open the `Modal`
1. Interact with the `Popover` activator.

## Reduced test case

https://codesandbox.io/s/p7km9yv2vx . You may need to expand the browser viewport so that the `Modal` renders in the "desktop" version (centered in the middle of the screen). The mobile version tends to have enough space, and the popover works there.

## Specifications

- Are you using the React components? (Y/N): Y
- Polaris version number: 2.12.1
- Browser: Chrome 70.0.3538.77 (64 bit)
- Device: MacBook Pro
- Operating System: macOS Mojave 10.14.1

### PR DESCRIPTION
**Pull request by JoelBesada**
### WHY are these changes introduced?
(First Pull Request, here we go!)

Resolves #544.

It seemed like the implementation of `intersectionWithViewport` within the `Popover` component returned some (what I would assume to be) unintended values in certain cases. In the case reported in the issue above, a `Rect` with a negative height is returned. I changed the implementation so this shouldn't happen anymore. 

I couldn't find any more context around this however, since the commit that added this function points to a PR that is not available here. I can only assume the expected behaviour based on the name of the function, but since I don't know the full intention of the original implementation I might be breaking something else here that's not caught by the tests. Please let me know! 

### WHAT is this pull request doing?

* Changes the `intersectionWithViewport` implementation to correctly clip the given rect
* Adds a couple of test cases for `intersectionWithViewport`

### How to 🎩

🖥 [Local development instructions](https://github.com/Shopify/polaris-react/blob/master/README.md#development)
🗒 [General tophatting guidelines](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting.md)
📄 [Changelog guidelines](https://github.com/Shopify/polaris-react/blob/master/documentation/Versioning%20and%20changelog.md)

You should be able to open and close the popover within the modal in the Playground example.

<details>
<summary>Copy-paste this code in <code>playground/Playground.tsx</code>:</summary>

```jsx
import * as React from 'react';
import {Page, ActionList, Button, Popover, Modal} from '@shopify/polaris';
import {noop} from '@shopify/javascript-utilities/other';

interface State {
  active: boolean;
}

export default class Playground extends React.Component<never, State> {
  state = {
    active: true,
  };

  togglePopover = () => {
    this.setState(({active}) => {
      return {active: !active};
    });
  };

  render() {
    const activator = (
      <Button onClick={this.togglePopover}>More actions</Button>
    );

    return (
      <Page title="Playground">
        <Modal title="Testing" open onClose={noop}>
          <Modal.Section>
            <Popover
              active={this.state.active}
              activator={activator}
              onClose={this.togglePopover}
            >
              <ActionList items={[{content: 'Import'}, {content: 'Export'}]} />
            </Popover>
          </Modal.Section>
        </Modal>
      </Page>
    );
  }
}
```

</details>

### 🎩 checklist

* [x] Tested on [mobile](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting.md#cross-browser-testing)
* [x] Tested on [multiple browsers](https://help.shopify.com/en/manual/intro-to-shopify/shopify-admin/supported-browsers)
* [x] Tested for [accessibility](https://github.com/Shopify/polaris-react/blob/master/documentation/Accessibility%20testing.md)
* [x] Updated [UNRELEASED.md](https://github.com/Shopify/polaris-react/blob/master/UNRELEASED.md)

### CODE REVIEW

**Comment by danrosenthal** [comment]
Thanks so much for the contribution @JoelBesada! We'll give this a look through and get back to you

File: UNRELEASED.md [hunk 1]
```diff
@@ -12,6 +12,8 @@ Use [the changelog guidelines](https://git.io/polaris-changelog-guidelines) to f
 
 ### Bug fixes
 
+- Fix Popover not opening in a small Scrollable container [#658](https://github.com/Shopify/polaris-react/pull/658)
```
Comment (danrosenthal) [inline UNRELEASED.md]:
```suggestion
- Fixed Popover not opening in a small Scrollable container ([#658](https://github.com/Shopify/polaris-react/pull/658))
```

**Comment by tmlayton** [comment]
And the 🎩 looks good, thanks for this fix!

**Comment by AndrewMusgrave** [comment]
Re-pinged reviewers for a fresh set of 👀

File: src/components/PositionedOverlay/PositionedOverlay.tsx [hunk 2]
```diff
@@ -248,22 +248,20 @@ export default class PositionedOverlay extends React.PureComponent<
...
-      rect.width - rect.left + viewport.left,
-      viewport.width,
-      viewport.width - rect.left,
-    ),
+    top,
+    left,
+    height: bottom - top,
+    width: right - left,
   });
 }
```
Comment (danrosenthal) [inline src/components/PositionedOverlay/PositionedOverlay.tsx]: This change makes the code so much more readable 👍

**Comment by AndrewMusgrave** [comment]
Looks like this is ready for review pinging reviewers again. cc/ @tmlayton @dleroux @solonaarmstrong

**Comment by danrosenthal** [comment]
I'm going to rebase this and merge, unless any pinged reviewers have any objections

**Comment by danrosenthal** [comment]
@JoelBesada would you mind resolving this conflict so I can go ahead and merge?

(unless any pinged reviewers have objections)

**Comment by JoelBesada** [comment]
@danrosenthal resolved

**Comment by danrosenthal** [comment]
Thanks! I'll merge on 🍏

**Pull request merged**